content/posts/2023-12-optimising-spark-memory-usage.md

```diff
@@ -0,0 +1,278 @@
+has more than two million weekly downloads and is also used by
+[Matteo Collina](https://github.com/mcollina) in his tools like
+[autocannon](https://www.npmjs.com/package/autocannon). I highly respect Matteo for his knowledge of
+Node.js & V8 performance tuning. If `hdr-histogram-js` is good for him, then it surely should be
+good for us, too. Right?
+
+Well, our data set is different from what autocannon produces. I _think_ the major difference is
+that we have a much wider range of possible values. For example, "CAR size in bytes" can be anywhere
+between zero to hundreds of million.
+
+Anyway, this is what I saw in our logs:
+
+- For many rounds, the histogram module was throwing an error while adding more data points.
+- Often, the values at percentiles were inaccurate. For example, "CAR size in bytes" at p5 and p10
+  was zero.
+
+Instead of debugging the library, I decided to calculate the histogram ourselves. If we can fit all
+measurements in memory, we can surely keep in memory an array of all values for an individual
+measurement field like "CAR size in bytes" or "TTFB", too.
+
+Calculating values at a given percentile is easy when we have all data points in memory:
+
+1. Sort the array by values.
+
+   _Remember to provide a custom comparator function because JavaScript sorts lexicographically by
+   default: `10` comes before `2`!_
+
+2. Calculate the value at the percentile `p` as the array item at position
+   `Math.ceil(p * count / 100)`.
+
+This formula produces the so-called _discrete percentile_. As I quickly discovered during testing,
+discrete percentiles are not very useful if you have a small number of data points. For example, if
+you record values `[1, 50, 100]`, the discrete percentile 5% (p5) is the same as the median value
+(p50): `50`.
+
+Fortunately, there are also _continuous percentiles_ that use interpolation to get finer-grained
+values. In my previous example, p5 is calculated as `5.9`.
+
+Here is the final implementation we use:
+
+```js
+export const getValueAtPercentile = (values, p) => {
+  // See https://www.calculatorsoup.com/calculators/statistics/percentile-calculator.php
+  // 1. Arrange n number of data points in ascending order: x1, x2, x3, ... xn
+  const n = values.length;
+
+  // 2. Calculate the rank r for the percentile p you want to find: r = (p/100) * (n - 1) + 1
+  const r = ((n - 1) * p) / 100 + 1;
+
+  // 3. If r is an integer then the data value at location r, xr, is the percentile p: p = xr
+  if (Number.isInteger(r)) {
+    // Remember that we are indexing from 0, i.e. x1 is stored in values[0]
+    return values[r - 1];
+  }
+
+  const ri = Math.floor(r);
+  const rf = r - ri;
+
+  // 4. If r is not an integer, p is interpolated using ri, the integer part of r,
+  // and rf, the fractional part of r:
+  // p = xri + rf * (xri+1 - xri)
+  return values[ri - 1] + rf * (values[ri] - values[ri - 1]);
```

Comment:
Is it worth extracting this code into a small npm package? WDYT, @juliangruber?

Reply:
Would be nice yeah! Also lets the reader easily reuse the code

Reply:
Found an existing package that already does this: https://www.npmjs.com/package/just-percentile

It makes me wonder why I didn't find this npm package before!

Reply:
Added another paragraph at the end of this section:

```md
Note: While revisiting this blog post, I performed another search on npmjs.com and found a package
that provides the same functionality:
[just-percentile](https://www.npmjs.com/package/just-percentile). I'll use it the next time I need
to calculate percentiles.
```